pkg/ibmvpc/parser.go

```diff
@@ -557,90 +585,95 @@ func getFipConfig(
+			VPCResource: vpcmodel.VPCResource{ResourceName: vpcName, ResourceUID: *vpc.CRN, ResourceType: ResourceTypeVPC},
 			nodes:       []vpcmodel.Node{},
 			zones:       map[string]*Zone{},
 		}
-		res.NodeSets = append(res.NodeSets, vpcNodeSet)
-		res.NameToResource[vpcNodeSet.Name()] = vpcNodeSet
-		// assuming one vpc
-		return vpcNodeSet, nil
+		res[vpcName] = NewEmptyCloudConfig()
+		res[vpcName].NodeSets = append(res[vpcName].NodeSets, vpcNodeSet)
+		res[vpcName].NameToResource[vpcNodeSet.Name()] = vpcNodeSet
 	}
-	return nil, errors.New("could not find a VPC config")
+	if len(res) == 0 {
+		return errors.New("could not find any VPC to analyze")
+	}
+	return nil
 }
 
-func singleVPCErr(vpcName, layerVPCName string) error {
-	return fmt.Errorf("NACL/SG VPC (%s) is different from its layer's VPC (%s)", vpcName, layerVPCName)
+func parseSGTargets(sgResource *SecurityGroup,
+	sg *vpc1.SecurityGroup,
+	res map[string]*vpcmodel.CloudConfig,
+	intfNameToIntf map[string]*NetworkInterface,
+	vpcName string) {
+	targets := sg.Targets // *SecurityGroupTargetReference
+	// type SecurityGroupTargetReference struct
+	for _, target := range targets {
+		if targetIntfRef, ok := target.(*vpc1.SecurityGroupTargetReference); ok {
+			// get from target name + resource type -> find the address of the target
+			targetType := *targetIntfRef.ResourceType
+			targetName := *targetIntfRef.Name
+			if targetType == networkInterfaceResourceType {
+				if intfNode, ok := intfNameToIntf[targetName]; ok {
+					sgResource.members[intfNode.address] = intfNode
+				}
+			} else if targetType == vpeResourceType {
+				if vpe, ok := res[vpcName].NameToResource[targetName]; ok {
+					vpeObj := vpe.(*Vpe)
+					for _, n := range vpeObj.nodes {
+						nIP := n.(*ReservedIP)
+						sgResource.members[nIP.address] = n
+					}
+				}
+			}
+		}
+	}
 }
 
 func getSGconfig(rc *ResourcesContainer,
-	res *vpcmodel.CloudConfig,
+	res map[string]*vpcmodel.CloudConfig,
 	intfNameToIntf map[string]*NetworkInterface,
 	skipByVPC func(string) bool) error {
 	sgMap := map[string]*SecurityGroup{}
-	sgList := []*SecurityGroup{}
-	var layerVPC *VPC
-	for i := range rc.sgList {
-		sg := rc.sgList[i]
+	sgList := map[string][]*SecurityGroup{}
+	for vpcName := range res {
+		sgList[vpcName] = []*SecurityGroup{}
+	}
+	for _, sg := range rc.sgList {
 		if skipByVPC(*sg.VPC.CRN) {
 			continue
 		}
-		vpc, err := getVPCObjectByName(res, *sg.VPC.Name)
+		vpcName := *sg.VPC.Name
+		vpc, err := getVPCObjectByName(res, vpcName)
 		if err != nil {
 			return err
 		}
-		if layerVPC == nil {
-			layerVPC = vpc
-		} else if layerVPC.Name() != vpc.Name() {
-			return singleVPCErr(vpc.Name(), layerVPC.Name())
-		}
 
 		sgResource := &SecurityGroup{
 			VPCResource: vpcmodel.VPCResource{
 				ResourceName: *sg.Name, ResourceUID: *sg.CRN, ResourceType: ResourceTypeSG},
 			analyzer: NewSGAnalyzer(sg), members: map[string]vpcmodel.Node{}, vpc: vpc,
 		}
 		sgMap[*sg.Name] = sgResource
```

Comment:
Isn't this problematic? Different Security Groups with the same name may live in different VPCs

Reply:
right, fixed (`sgMap` is now per each VPC)